Add debug-cudf-pandas skill for fixing pandas test compatibility failures (#22625)
rapidsai/cudf: .agents/skills/debug-cudf-pandas/SKILL.md

---
name: debug-cudf-pandas
description: Debug and fix pandas test suite failures under the cudf.pandas compatibility layer. Use when given pytest node IDs of failing pandas tests that need to be fixed for cudf.pandas compatibility.
---

# Debug cudf.pandas Compatibility Failures

## Overview

`cudf.pandas` is a zero-code-change GPU accelerator for pandas. It intercepts `import pandas` via a Python `MetaPathFinder` (`python/cudf/cudf/pandas/module_accelerator.py`) and wraps all pandas objects and functions in proxy types that try the cudf (GPU) path first, falling back to pandas (CPU) if cudf raises an exception.

> When the pandas test suite is run with `-p cudf.pandas`, test failures indicate one of:

Reviewer: Not sure if you want to call out rarer forms of failures

* A bug spcifically when converting from a cuDF object to a pandas object and vice versa (`to/from_pandas`)
* A bug or a modifications in the testing setup scripts
* A bug in pandas

Author: Done — added to/from_pandas conversion bugs, test setup bugs, and pandas bugs to the failure mode list.

---

## Step 2 — Understand the Test

After the first run, the test file exists at:

```
pandas-testing/pandas-tests/tests/<module>/<test_file>.py
```

Read the test. Identify:
- Which pandas API is being exercised (method name, class, module)
- What the test asserts — this is the exact behavior cudf must match
- Whether parameters narrow the scope (e.g. only `Float64` dtype, only `skipna=False`)

The assertion error output from Step 1 tells you what cudf produced vs. what was expected. Use both together to understand the gap.

---

## Step 3 — Diagnose Root Cause

**Always check cudf's direct behavior first.** Do not jump to proxy investigation.

### 3a. Direct cudf test (most important step)

Write a minimal script that exercises the API using cudf directly, without going through `cudf.pandas`:

```python
import cudf
import pandas as pd

# Mirror the test setup
pdf = pd.DataFrame({"value": [1.0, None, 2.0, 3.0]}, dtype="Float64")
cdf = cudf.from_pandas(pdf)

pd_result = pdf.groupby(level=0).first(skipna=False)
cudf_result = cdf.groupby(level=0).first(skipna=False)

print("pandas:", pd_result)
print("cudf:  ", cudf_result)

> print("match:", pd_result.equals(cudf_result.to_pandas()))

Reviewer: Not sure how much an agent will knows this is strictly an example (and not always a valid way to make a comparison).

Also there are forms of tests that check exceptions are raised that might structure a comparison script differently

Author: Good point. Added explicit language that this is "strictly an example" and to adapt the structure for what the test does. Also added a note about structuring scripts differently for exception-testing cases.

```

Save to a temporary file (e.g. `test_debug.py`) and run:

```bash
python test_debug.py
```

Results:
- **cudf result differs from pandas** → cudf implementation bug → go to Step 4a

> - **cudf raises `NotImplementedError`** → missing feature → evaluate scope; may need user input if the feature is large

Reviewer: ```suggestion
- **cudf raises an exception** → missing feature → evaluate scope; may need user input if the feature is large
```
Also this may be OK if the test is testing for an exception to raise

Author: Done — changed to "raises an exception" and added a note that this may be OK if the test is verifying an exception should be raised.

- **cudf result matches pandas** → proxy/dispatch bug → go to Step 4b

### 3b. Environment variable diagnostics (run through cudf.pandas)

Use these env vars to trace what is happening at the proxy layer:

| Variable | Effect |
|----------|--------|
| `CUDF_PANDAS_FAIL_ON_FALLBACK=1` | Raises instead of silently falling back — shows exactly which operation triggers fallback |
| `LOG_FAST_FALLBACK=1` | Logs every fallback with function name and exception |
| `CUDF_PANDAS_DEBUGGING=1` | Runs both cudf and pandas paths in parallel, warns on result divergence |

```bash
rm -rf pandas-testing/pandas-tests/
CUDF_PANDAS_FAIL_ON_FALLBACK=1 bash python/cudf/cudf/pandas/scripts/run-pandas-tests.sh \
    "<node_id>" -xvs
```

```bash
rm -rf pandas-testing/pandas-tests/
LOG_FAST_FALLBACK=1 bash python/cudf/cudf/pandas/scripts/run-pandas-tests.sh \
    "<node_id>" -xvs 2>&1 | grep -i fallback
```

> ### 3c. Standalone instrumented script
> 
> For deeper investigation, copy the relevant test body into a temporary script (e.g. `test_debug.py`), add print statements or assertions at intermediate steps, then run through the proxy layer:
> 
> ```bash
> python -m cudf.pandas test_debug.py
> ```
> 
> This gives you full control to narrow down exactly where the divergence begins.

Reviewer 2: ```suggestion
### 3c. Standalone instrumented script

For deeper investigation, copy the relevant test body into a temporary script (e.g. `test_debug.py`), add print statements or assertions at intermediate steps, then run through the proxy layer or try to write matching pandas code to see the differences in behaviors and fix them:

```bash
python -m cudf.pandas test_debug.py
```

This gives you full control to narrow down exactly where the divergence begins.

```

Author: Done — expanded to mention writing matching pandas code to compare behaviors.

---

> ## Step 4a — Fix a cudf Implementation Bug

Reviewer: I think it's also important to make the agent evaluate the cause of the failure and classify if the failure mode and a potential fix is specific (e.g. 1 method handles a keyword incorrectly) vs a broad failure mode (e.g. dtype casting inconsistency)

Author: Added a paragraph at the top of Step 4a requiring the agent to classify the failure mode (specific vs broad) before writing the fix, and to consider applying broad fixes at a shared/base level.

> The fix must be **minimal and tightly scoped**. Fixing one edge case can break another test. Do not refactor surrounding code while fixing, and do not add `mode.pandas_compatible` guards (ask the user first).

Reviewer 2: ```suggestion
The fix must be **tightly scoped** and try to keep it minimal. Fixing one edge case can break another test. Do not refactor surrounding code while fixing, and do not add `mode.pandas_compatible` guards (ask the user first).
```

Author: Done — now says "tightly scoped and try to keep it minimal."

| Series method | `python/cudf/cudf/core/series.py` |
| Index operation | `python/cudf/cudf/core/index.py` |
| Shared DataFrame/Series method | `python/cudf/cudf/core/indexed_frame.py` |
| Column-level operation | `python/cudf/cudf/core/column/*.py` |
| GroupBy operation | `python/cudf/cudf/core/groupby/` |
| IO operation | `python/cudf/cudf/io/` |

When a fix touches column operations that call into `pylibcudf`, consult `python/CLAUDE.md` for the required buffer access pattern (`column.access(mode="read", scope="internal")`).

Note: `mode.pandas_compatible` is automatically set to `True` when cudf.pandas is active. Account for this in any conditional logic, but do not add new guards for it without explicit user approval.

---

## Step 4b — Fix a Proxy/Dispatch Bug

Only reach this step after Step 3a has confirmed that cudf itself is correct.

**Most common cause**: a pandas return type has no registered cudf proxy. Check `python/cudf/cudf/pandas/_wrappers/pandas.py` — this file registers which pandas types map to which cudf types using `make_final_proxy_type()` and related functions. If a new pandas type needs wrapping, add the registration here.

> **Do not modify** `python/cudf/cudf/pandas/fast_slow_proxy.py` or `python/cudf/cudf/pandas/module_accelerator.py`. These are core infrastructure files. If you believe the bug is in one of them, **stop and report your findings to the user** — do not attempt to fix them.

Reviewer 2: ```suggestion
**Do not modify** `python/cudf/cudf/pandas/fast_slow_proxy.py` or `python/cudf/cudf/pandas/module_accelerator.py`. These are core infrastructure files. Fix them only if you believe the bug is in one of them.
```

Trying to reduce the time delays here. Making agents wait for us devs input can be reduced.

Author: Done — changed to: "`fast_slow_proxy.py` and `module_accelerator.py` are core infrastructure files. Fix them only if you believe the bug is in one of them."

**a. Target test passes:**

```bash
rm -rf pandas-testing/pandas-tests/
bash python/cudf/cudf/pandas/scripts/run-pandas-tests.sh "<node_id>" -xvs
```

Expected: exit code 0, test shows `PASSED`.

**b. Fix runs on GPU (no silent fallback):**

```bash
rm -rf pandas-testing/pandas-tests/
CUDF_PANDAS_FAIL_ON_FALLBACK=1 bash python/cudf/cudf/pandas/scripts/run-pandas-tests.sh \
    "<node_id>" -xvs
```

Expected: test still passes. If this fails, the fix works by falling back to pandas rather than actually fixing cudf — that is not acceptable.

Exception: some tests intentionally validate fallback behavior. If `FAIL_ON_FALLBACK` causes this test to fail but the test logic requires fallback, skip this check for that specific test and note the justification.

**c. No regressions in the module:**

```bash
rm -rf pandas-testing/pandas-tests/
bash python/cudf/cudf/pandas/scripts/run-pandas-tests.sh \
    "tests/<module_directory>/" --tb=line -q
```

Replace `<module_directory>` with the directory containing your test (e.g. `tests/groupby/`). Any new failures that are not already listed in `conftest-patch.py` must be investigated before committing.

---

> ## Step 6 — Commit

Reviewer 2: Can we add another step before commit that will be something like:

```
Step 7 - Add unit tests in `cudf` classic or `cudf.pandas` where appropriate depending on the fix. While adding unit tests, see if you can add to existing test parameterizations and if none present then write a test that only compares to pandas (with something like `assert_eq` or `assert_exceptions_equal`)

Author: Added as Step 5d (before commit). Includes guidance to prefer adding to existing test parameterizations, and to use `assert_eq` or `assert_exceptions_equal` for new standalone tests.

git add python/cudf/cudf/pandas/scripts/conftest-patch.py   # xfail removal
git status                          # verify nothing from pandas-testing/ is staged
git commit -m "fix(cudf.pandas): <short description>

Fixes the following failing pandas tests:
- tests/groupby/test_reductions.py::test_first_last_skipna[Float64-False-False-first]"
```

If many node IDs share the same root cause, summarize them in the commit message rather than listing every parametrized variant. Commit message should convey what was wrong and what was fixed.

---

## STOP Conditions

Stop immediately and report findings to the user when any of the following apply:

- The fix requires modifying `.pyx`, `.cu`, `.cuh`, or `CMakeLists.txt` files — this requires compilation, which is outside the scope of this skill
- The fix requires modifying `fast_slow_proxy.py` or `module_accelerator.py` — core proxy infrastructure, requires human judgment
- Investigation reveals the behavioral divergence is **intentional** — cudf was deliberately designed to differ from pandas
- The fix requires implementing an entirely new method or feature from scratch, not merely correcting a bug in existing code

> - You have been through 3 or more investigation cycles without converging on a fix

Reviewer 2: ```suggestion
- Investigation reveals the behavioral divergence is **intentional** — cudf was deliberately designed to differ from pandas
- You have been through 3 or more investigation cycles without converging on a fix
```

I think we should allow these. We did make a ton progress and bug findings in fast slow proxy mechanisms.

Author: Agreed — removed both STOP conditions. The `fast_slow_proxy.py`/`module_accelerator.py` restriction is now relaxed to "Fix them only if you believe the bug is in one of them." And the "new feature" restriction is gone entirely.

- The fix would add a new `mode.pandas_compatible` conditional

For intentional divergence: stop and ask the user. In most cases, the goal is to make cudf match pandas. Only when the divergence has significant performance implications should `mode.pandas_compatible` be used — and even then, sparingly. cudf already overuses this flag; the default behavior should agree with pandas.

---

## Important Notes

- `mode.pandas_compatible` is automatically set to `True` when `cudf.pandas` is active. This is done at the end of `python/cudf/cudf/pandas/_wrappers/pandas.py`.

> - cudf Python is almost entirely pure Python — changes to `.py` files take effect immediately without rebuilding.

Reviewer 2: This is only true for inplace installs(which is not the case for `./build.sh`), claude figured it out and worked around it by either making inplace installs or copying to site-packages.

Author: Fixed — now says "For **inplace installs** (e.g. `pip install -e .`), changes to `.py` files take effect immediately. For non-inplace installs (e.g. `./build.sh`), you must either reinstall or copy changed files to site-packages."

- `pandas-testing/pandas-tests/` must be deleted before each test run to avoid duplicate conftest hook registrations (the script appends the patch file on every run).
- Keys in all three `conftest-patch.py` dictionaries must remain in alphabetical order.
- Never write comments that explain what old code was replaced — write comments about what the code does and why.

> - Never modify the pandas test files themselves — fix cudf, not pandas.

Reviewer 2: ```suggestion
- Never modify the pandas test files themselves — fix cudf, not pandas.
- Never fix the testing APIs likes `assert_...`, fix the actual APIs
- First see if the problem is in `cudf` classic and fix it, if not then move over to `cudf.pandas`
```

Author: Done — added both to the Important Notes section:
- Never fix the testing APIs (like `assert_frame_equal`, `assert_series_equal`) — fix the actual APIs that produce wrong results.
- First see if the problem is in cudf classic and fix it there; if not, then move over to cudf.pandas.
